lib/ecto/repo/preloader.ex

```diff
@@ -154,6 +155,67 @@ defmodule Ecto.Repo.Preloader do
 
   defp preload_assocs([], [], _repo_name, _tuplet), do: []
 
+  defp preload_embeds(structs, [], _repo_name, _tuplet), do: structs
+
+  defp preload_embeds(structs, embeds, repo_name, tuplet) do
+    # prepare all embeds of the same owner field into a list
+    embed_tuples = Enum.reduce(
+      embeds, [],
+      fn {embed, sub_preloads}, embeds_acc ->
+        %{field: field, cardinality: card} = embed
+
+        all_schemas = Enum.map(structs, &Map.get(&1, field))
+        case card do
+          :many ->
+            list_of_lists = all_schemas
+            # list with quantity of items in embeds :many per owner field
+            track_positions = Enum.map(list_of_lists, &Enum.count(&1))
+            # unwrap the lists
+            flat_schemas = List.flatten(list_of_lists)
+            [{flat_schemas, embed, sub_preloads, track_positions } | embeds_acc ]
+          :one ->
+            list_of_schemas = all_schemas
+            [{list_of_schemas, embed, sub_preloads, :one } | embeds_acc ]
+        end
+    end)
```

Comment:
Oh look at you, rewriting the entire thing in one go and without mistakes. Awesome @greg-rychlewski , How do you do that? This is like 50% less code! 

I will feel a little guilty putting my name on the coming commit.

I like the `foo([head | rest]) do: .... foo(rest)` recursion, much concise. Also the usage of `flat_map_reduce`, very tidy.

Reply:
> I like the foo([head | rest]) do: .... foo(rest) recursion, much concise. Also the usage of flat_map_reduce, very tidy.

Oh yes those are nice tricks you get used to after some time using this language =).

Btw I updated the suggestion just a bit because you made me realize that `nil` can be passed in the list of structs. I didn't think it was possible before so this new suggestion is more fool proof.